### PR TITLE
[FIX] hr_holidays: correctly calculate leave duration

### PR DESCRIPTION
**Pull request by taqu-odoo**
Steps
-----
- Time off installed
- Put an employee on a different schedule than the company's
default one and make him for example start earlier.
- Create a new Time Off type, set "Requires allocation" to
No Limit.
- Approvals > Time off > New and create a time off (mode "by
company") on the day where the employee is starting earlier.
- Approve it.
- Approvals > Time off : the leave's duration for the employee
on a different schedule is incorrect (slightly less than 1 day).

Issue
-----
When creating a leave on a company level, a leave will first be
created with start and end hours based on the company's default
schedule.
When approving the leave, a leave is then created for each
affected employee. The date_from and date_to received by
create are based on the company's default schedule. This is not
generally problematic as it is taken into account. For example,
the date_from and date_to of the employee's leave are correct
even if he is on a different schedule.

However, when we compute the number of days, it is not accounted for.
https://github.com/odoo/odoo/blob/64cbe389e698398eee93ebde9c61b2ee79756380/addons/hr_holidays/models/hr_leave.py#L918
https://github.com/odoo/odoo/blob/64cbe389e698398eee93ebde9c61b2ee79756380/addons/hr_holidays/models/hr_leave.py#L744
As a result, the hours worked by an employee on a different schedule
which are "outside" of the company's default schedule don't count
towards the calculation.
https://github.com/odoo/odoo/blob/3eec4d6a69ff76cbf750b960dc8db2eb5e0a45f9/addons/resource/models/resource_mixin.py#L94-101

Note: the issue is the same for leaves given to a department or an
employee tag since they are created the same way.

Fix
-----
When creating a leave for employees, if the leave is not for half
days or hours, give the whole day as a leave to account for different
schedules. Note that the actual date_from and date_to of each employee's
leave are still based on their actual attendance, but if an employee
has hours outside of the company's schedule, they will count towards
the leave's duration.

opw-3703793

### CODE REVIEW

**Comment by taqu-odoo** [comment]
@fw-bot up to saas-16.4 https://github.com/odoo/odoo/commit/6074d86b2076117978136259d70d043ef3dba1ee refactored the leave duration in 17.0 and fixed this issue.

**Review by nle-odoo — APPROVED** [review]
seems to make sense, but I don't know if there is better ways to do it

seeing the code below does something similar:

https://github.com/odoo/odoo/blob/b4c8625389bf4e23a64a893c1d7cadb893c6a591/addons/hr_holidays/models/hr_leave.py#L922-L934

I guess it's necessary.
    )
    
```

but it doesn't remove much so current version looks good

File: addons/hr_holidays/models/hr_leave.py [hunk 1]
```diff
@@ -915,7 +915,16 @@ def create(self, vals_list):
         if self._context.get('leave_compute_date_from_to') and employees:
             employee_leave_date_duration = defaultdict(dict)
             for (date_from, date_to), employee_ids in leave_date_employees.items():
-                employee_leave_date_duration[(date_from, date_to)] = self._get_number_of_days_batch(date_from, date_to, employee_ids)
+                for calendar in employees.resource_calendar_id:
+                    calendar_employees = employees.filtered(lambda emp: emp.resource_calendar_id == calendar)
+                    attendance_from, attendance_to = self._get_attendances(calendar_employees, date_from.date(), date_to.date())
```
[the remark this answers is not in the document]
Comment (taqu-odoo) [reply addons/hr_holidays/models/hr_leave.py:920]: In this case I think we want the start and end hours of the attendance for that day, so we give only the days as arguments

**Comment by taqu-odoo** [comment]
It seems to me we could also calculate the number of days of a leave by passing to _get_number_of_days_batch the start of the first day and the end of the last day, but not too sure about that.

**Review by Bertrand2 — COMMENTED** [review]
What happens if the leave is for half a day/specific hours? Won't that create a whole day leave when creating the individual leaves?

**Comment by taqu-odoo** [comment]
I've modified the PR. I think conceptually it makes sense to consider the whole day as a leave if we have different schedules, let me know what you think.

**Comment by Bertrand2** [comment]
This leaves the initial issue for half-days :thinking: Isn't there any way to just compute date_from and date_to based from the UX fields?

**Comment by Bertrand2** [comment]
This doesn't work for half_days. The leave will indeed have a 0.5 days duration, but the date_from and date_to will cover the whole day.

**Comment by taqu-odoo** [comment]
Hello @Bertrand2, date_from and date_to are now correct for leaves in half_days or in hours.

**Comment by taqu-odoo** [comment]
Thanks for all the suggestions :+1: Changes made.

File: addons/hr_holidays/tests/test_company_leave.py [hunk 2]
```diff
@@ -334,3 +334,103 @@ def test_leave_whole_company_07(self):
...
+        calendar = self.env['resource.calendar'].create({
+            'name': 'Different schedule',
+            'attendance_ids': [(5, 0, 0),
+                               (0, 0, {
+                                   'name': 'monday morning, earlier start',
+                                   'hour_from': 7.5,
+                                   'hour_to': 12,
+                                   'day_period': 'morning',
+                                   'dayofweek': '0',
+                               }),
```
[the remark this answers is not in the document]
Comment (Bertrand2) [reply addons/hr_holidays/tests/test_company_leave.py]: This wasn't resolved according to my comment

File: addons/hr_holidays/tests/test_company_leave.py [hunk 3]
```diff
@@ -334,3 +334,104 @@ def test_leave_whole_company_07(self):
...
+                                   'hour_from': 10,
+                                   'hour_to': 12,
+                                   'day_period': 'morning',
+                                   'dayofweek': '0',
+                               })]
+        })
+        self.employee.resource_calendar_id = calendar
+
+        # employee on default schedule
+        self.employee2 = self.env['hr.employee'].create({
```
Comment (Bertrand2) [inline addons/hr_holidays/tests/test_company_leave.py:364]: This shouldn't write on self
```suggestion
        employee2 = self.env['hr.employee'].create({
```

**Review by Bertrand2 — COMMENTED** [review]
Just one last precision over what I meant for the tests, otherwise LGTM

**Comment by taqu-odoo** [comment]
Changed :slightly_smiling_face: Thank you for the help

**Review by Bertrand2 — APPROVED** [review]
LGTM! Thanks

**Comment by Bertrand2** [comment]
robodoo r+

**Pull request closed without merging**